### DISCUSSION TITLE
Exporting from inlet to Kafka errors

### DISCUSSION THREAD

**Discussion by GeorgeMaged** [akvorado/akvorado#1239] [Q&A]
I am trying to deploy akvorado  using docker compose , all containers are running. I have Cisco router netflow-configured to send flows to port 7779 and edited that in docker-compose.yml for the inlet service.It successfully listens to the flows sent but nothing exposed to Kafka or clickhouse DB. These are the logs from this command: 

 curl -s http://akvorado/api/v0/inlet/metrics | grep '^akvorado_inlet' | grep _errors

        Line 55: # HELP akvorado_inlet_core_flows_errors_total Number of flows with errors.
	Line 56: # TYPE akvorado_inlet_core_flows_errors_total counter
	Line 57: akvorado_inlet_core_flows_errors_total{error="SNMP cache miss",exporter="172.19.0.1"} 9.170946e+06
	Line 58: akvorado_inlet_core_flows_errors_total{error="sampling rate missing",exporter="172.19.0.1"} 985599
	Line 65: # HELP akvorado_inlet_flow_decoder_errors_total Decoder processed error count.
	Line 66: # TYPE akvorado_inlet_flow_decoder_errors_total counter
	Line 67: akvorado_inlet_flow_decoder_errors_total{name="netflow"} 365
	Line 71: # HELP akvorado_inlet_flow_decoder_netflow_errors_total Netflows processed errors.
	Line 72: # TYPE akvorado_inlet_flow_decoder_netflow_errors_total counter
	Line 73: akvorado_inlet_flow_decoder_netflow_errors_total{error="NetFlow v9 decoding error",exporter="172.19.0.1"} 365

Is there a workaround to have the flows without the SNMP polling?

**Comment by vincentbernat** [comment]
Yes, metadata can be configured statically and there is the possibility to have a default value. Check the documentation.

**Comment by GeorgeMaged** [comment]
I have another question regarding the fields sent in the Netflow v9 standard, There are fields like (vrf)or (mpls exp bits) and (interface number) that I am interested in but they are not collected or exported to Kafka but the exporter is sending these fields already. How to show them? Is there a way to enable these fields ?

**Reply by vincentbernat** [reply]
There are fields that can be enabled optionally. For the others, they need to be implemented specifically. PR are welcome.

**Comment by GeorgeMaged** [comment]
Hello @vincentbernat , I am trying to get familiar more with avkorado.
Currently trying to show the interface index as a filter in avkorado, Internally in the inlet.yaml I am adding in a static section different interface indexes for the incoming flows for testing and yes it is able to identify the indexes from the incoming flows. However, I cant find it as a metric to show/enable in the definition.go file so I will be able to use this in the GUI filter section.
Also, As I am running the solution in Docker, The exporter address because of the networking docker settings will be always the internal IP of the inlet container, Can there be a workaround for this to get the actual ExporterAddress that is sending the flows?

**Reply by vincentbernat** [reply]
Interface indexes are not exported in the flows. You can put them in the interface description with a classifier with `SetDescription()` as you have access to `Interface.Index` there.

You may need to disable Docker userland proxy for the second problem.

**Comment by GeorgeMaged** [comment]
When trying to use the SetDescription() classifer and passing the Interace.Index as an argument for it , this log error appeared from the orchestrator: 

Error: unable to parse configuration: 1 error(s) decoding:
                                                        2024-06-02T11:36:16.753248100Z 
                                                                                       2024-06-02T11:36:16.753264600Z * error decoding 'Inlet[0].Core.InterfaceClassifiers[0]': cannot compile interface classifier rule "SetDescription(Interface.Index)": cannot use uint32 as argument (type string) to call SetDescription  (1:26)
                                                              2024-06-02T11:36:16.753277600Z  | SetDescription(Interface.Index)
                                                                                                                               2024-06-02T11:36:16.753288300Z  | .........................^


This is because SetDescription() expects a string type while Interface.Index is uint type , which one of them to change its type from classifier.go ?

**Reply by vincentbernat** [reply]
See here: https://expr-lang.org/docs/language-definition#string

**Reply by GeorgeMaged** [reply]
Hello Vincent,
Currently I am in PoC for Akvoardo, trying with 1 Cisco router is configured to export the netflow v9 to my server which has the akvorado solution listening to port 2055 and successfully it receives the flows.
My first issue is that on some interfaces of the router which has a high throughput in peak hours
( e.g. 22Gbps for 5 minutes resolution) , I find when querying the database itself and aggregating for every 5 minutes the sun of bytes or calculations of bps , it maximum reaches like 16 or 17 Gbps. In addition, I checked the akvoardo_inlet_in_drops counter after increasing the number of workers and no significant losses exist so what are the possible causes for that as I want to be sure that it is reliable? 

Another thing regarding the GUI and graphs, I discovered that when selecting different time filters ( 1 hour , 6 hours, 24 hours ,7 days...etc) the query that gets the data differs in two things:
1)The table itself that gets the data from
2)The calculation of bps as it divides the sum of bytes *8*SamplingRate / (time) which differs per the time filter chosen. I want it to be always 5 minutes granularity and to always select from the flows table not other tables to be able to draw the reuired graphs , from which file can i edit in to have the required behaviour?

**Reply by vincentbernat** [reply]
On Cisco NCS, we have discovered that there was a bug in the sampling rate. It seems it should be a power of two. Maybe that's the same for you (moreover, changing the sampling also require to remove it from interfaces and adding it back).

For time division, there is nothing configurable. For the different tables, you can configure only the main table. Check `resolutions` here: https://demo.akvorado.net/docs/configuration#clickhouse.

**Reply by GeorgeMaged** [reply]
Actually I am giving the value of the sampling rate as 10000 but as a static value to to Akvorado as I am not using the snmp poller yet. 
That is actually the value configured on the router also, so What do I need to change here? 
What I meant for time division, That when I chose for example to see the bits per second for my traffic in the last hour the calculation used is that : 
(Sum(Bytes) * SamplingRate * 8) / **_18_** as xps , so on the graph it gives me value every 18 seconds which I am not interested to calculate the throughput for every 18 seconds. However, I need it to be for example to be divided by 300 ( 5 minutes granularity/resolution). That 18 seconds change if I increased the time filter for example to 6 hours or 12 hours ...etc 
So how to have this behavior and where to change the value of time it is divided by to get the xps ?

**Reply by vincentbernat** [reply]
For sampling rate, as I said, try to configure the router with a power of two (notably if this is a NCS).

As for the 18 seconds, this is not configurable. It is a computed value depending on the time range.

**Reply by GeorgeMaged** [reply]
![image](https://github.com/user-attachments/assets/1523d83b-e547-4b2e-9da1-a4772a472f19)

If it is a computed value, For example when I select the filter for 1 hour , it computes the step value by (18 s) for example and then when drawing the graph of throughput , it divides by this computed value which gives reading to the throughput every 18 seconds.
This is not the behavior or reading which I can depend on for my network analysis, As I want granularity of 5-minutes despite the the time range whether it is 1 hour or 1 week. Can you assist me in that for testing purposes as I want 5 minutes granularity to have readings every 5 minutes as as fixed value , in which files can I edit ?
